topologywatcher_http.go

```diff
@@ -0,0 +1,87 @@
+package gocbcorex
+
+import (
+	"context"
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+type TopologyWatcherHttpConfig struct {
+	HttpRoundTripper http.RoundTripper
+	Endpoints        []string
+	UserAgent        string
+	Authenticator    Authenticator
+	BucketName       string
+}
+
+type TopologyWatcherHttpOptions struct {
+	Logger *zap.Logger
+}
+
+type TopologyWatcherHttp struct {
+	logger *zap.Logger
+
+	configWatcher *ConfigWatcherHttp
+}
+
+func NewTopologyWatcherHttp(config *TopologyWatcherHttpConfig, opts *TopologyWatcherHttpOptions) (*TopologyWatcherHttp, error) {
+	configWatcher, err := NewConfigWatcherHttp(&ConfigWatcherHttpConfig{
+		HttpRoundTripper: config.HttpRoundTripper,
+		Endpoints:        config.Endpoints,
+		UserAgent:        config.UserAgent,
+		Authenticator:    config.Authenticator,
+		BucketName:       config.BucketName,
+	}, &ConfigWatcherHttpOptions{
+		Logger: opts.Logger,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := &TopologyWatcherHttp{
+		logger: opts.Logger,
+
+		configWatcher: configWatcher,
+	}
+
+	return watcher, nil
+}
+
+func (w *TopologyWatcherHttp) Watch(ctx context.Context) <-chan *ParsedConfig {
+	outCh := make(chan *ParsedConfig, 1)
+	go w.watchThread(ctx, outCh)
+	return outCh
+}
+
+func (w *TopologyWatcherHttp) Reconfigure(config *TopologyWatcherHttpConfig) error {
+	return w.configWatcher.Reconfigure(&ConfigWatcherHttpConfig{
+		HttpRoundTripper: config.HttpRoundTripper,
+		Endpoints:        config.Endpoints,
+		UserAgent:        config.UserAgent,
+		Authenticator:    config.Authenticator,
+		BucketName:       config.BucketName,
+	})
+}
+
+func (w *TopologyWatcherHttp) watchThread(ctx context.Context, outCh chan<- *ParsedConfig) {
+	var lastSentConfig *ParsedConfig
+	watchCh := w.configWatcher.Watch(ctx)
+
+	for {
+		var thisConfig *ParsedConfig
+		select {
+		case thisConfig = <-watchCh:
+		case <-ctx.Done():
+			close(outCh)
```

Comment:
In order to guarentee the cleanup of the HTTP client (and requests), shouldn't this instead be calling w.configWatcher.Close(), and then when watchCh closes we would close outCh?

Reply:
There is no Close, as it's all context driven - you're right though that closing outCh should occur after watchCh closes.